### ISSUE TITLE
How to zipper default-sparse arrays/domains with default-rectangular?

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#12424]
Pending some changes on master, these "ok to resort to serial" `[]` forall-loops do not compile, even though they can be executed as serial zippered loops:

* `[ (s,d) in zip(SPS,DR) ] ...` - causes a resolution error

* `[ (s,d) in zip(DR,SPS) ] ...` - says "can't zipper sparse with non-sparse"

where
 `SPS` - a default-sparse array/domain - or perhaps a LayoutCS one
 `DR` - a default-rectangular array/domain.

The cause is the compiler use of "resolve call" vs. "resolve function" :

* **resolve call** - determine which function to invoke for the given function call
  looking __only__ at function signatures
  including instantiating generic formals and evaluating where-clauses

* **resolve function** - resolve everything in the body of the function
  that was chosen during some "resolve call"

When looking at a zippered `[]` loop, the compiler takes these steps, assuming the all succeed:

* **resolve call** for the leader iterator
* **resolve function** for the leader iterator
* ... which is needed to determine the type of `followThis` for the next step
* **resolve call** for each of the followers
* (if the previous step fails for any of the followers, stop - convert to a serial loop instead)
* **resolve function** for each of the followers

In the `zip(SPS,DR)` case, **resolve call** for the DR follower succeeds -- because the signature is oblivious of the type of the `followThis` formal. However, **resolve function** for the DR follower fails -- because the body relies on `followThis` to be a tuple of ranges.

In the `zip(DR,SPS)` it goes like above except with the SPS follower. The catch-all follower is written to throw the "can't zipper sparse with non-sparse" compiler error.

**Comment by bradcray** [comment]
I don't think we should be supporting zippering of sparse domains/arrays with default rectangular by default.  My intuition is that to do so, the user may be required to call a `linearize()` iterator or method on one or both of the expressions to make them the same shape.

**Comment by vasslitvinov** [comment]
This question came up here:

#### Should we allow ANY two iterable things to be zippered together in a `[]`-loop?

Argument in support: Any two iterables can be zippered in a `for`-loop. A `[]`-loop falls back to a `for`-loop if it can't execute in parallel. So when we zipper any two things in a `[]`-loop, it should at least work as a `for`-loop.

Argument against: If we allow that, then we would be able to assign between any two arrays regardless of their shapes. Because `A1 = A2` is, semantically:
```chpl
[ (a1,a2) in zip(A1,A2) ] a1 = a2;
```
The following code would be legal:
```chpl
var A: [1..3, 1..5] real;
var B: [1..5, 1..3] real;
var C: [1..15] real;

A = B;
B = C;
A = C;
```
by silently falling back onto the respective arrays' serial iterators. This is not desirable.

Is there an argument for permitting the (sparse, rectangular) case that also keeps these cases illegal?

**Comment by vasslitvinov** [comment]
See also #11504
